# Class 4 - Loops (`while` and `for`)

Loops are repeated series of code. There are 2 kinds of loops - `while` and `for`.
When should we use which?:
- a pre-set number of repetitions is completed - `for`
- a condition is met - `while`

In [7]:
# WHILE
# Count to N
import time

n = int(input())
i = 0
while i < n:
    i = i + 1
    print(i, end = " ")
    time.sleep(.3)

 10


1 2 3 4 5 6 7 8 9 10 

In [3]:
# FOR
# Count to N
n = int(input())
for i in range(1, n+1):
    print(i, end = " ")
    time.sleep(.3)

 15


1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 

## `while`

In [16]:
# Count to N
from time import sleep

n = int(input())
i = 0
while i < n:
    sleep(1)
    print(i)
    i += 1    # i = i + 1
    
print('Goodbye')

 5


0
1
2
3
4
Goodbye


Small note on alternative Python syntax for item incrementing:

In [ ]:
i = 0
print(i)

In [ ]:
i += 1  # i = i + 1
print(i)

### Exercises
Write a loop that counts down from any input number (N > 1) to 1.

In [12]:
n = int(input()) # Assume that the number larger than 1

while n >= 1:
    print(n)
    n = n - 1
    sleep(.3)


 5


5
4
3
2
1


Write a loop that terminates once a correct 4-digit number has been enterred.

In [15]:
not(False)

True

In [13]:
password = 1234
user_input = int(input())

while not(user_input == password) :
    user_input = int(input())

print('Correct password')

 999
 1000
 1234


Correct password


Create a program that computes the average of a collection of values entered by the user. The user will enter 999 as a sentinel value to indicate
that no further values will be provided.

## `for`
`for` loops are used when a pre-set number of repetitions is needed

In [17]:
# Input and sum 3 numbers, using WHILE
sum_xs = 0
i = 0
while i < 3:
    x = int(input())
    sum_xs = sum_xs + x
    i = i + 1
print('Sum of inputs is', sum_xs)

 10
 100
 1000


Sum of inputs is 1110


In [18]:
# Input and sum 3 numbers, using FOR
sum_xs = 0
for i in range(3):
    x = int(input())
    sum_xs = sum_xs + x
print(sum_xs)

 10
 100
 1000


1110


In Python, `for` loops work on all iterables

In [2]:
from time import sleep

name = input('Enter your name:')

for x in name:
    print(x)
    sleep(.3)



Enter your name: david


d
a
v
i
d


### Using `range()` with `for`
`range()` is a very commonly used function together with `for` loops. So what is it?

In [27]:
x = range(20)
print(x)

range(0, 20)


In [26]:
list(range(20))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]

In [30]:
round(3.14, 1)

3.1

In [31]:
print(list(range(10)))
print(list(range(5, 10)))
print(list(range(0, 10, 3)))
list(range(10,0, -1))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[5, 6, 7, 8, 9]
[0, 3, 6, 9]


[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

In [32]:
print(type(range(5)))

<class 'range'>


In [33]:
27_541_113 in range(0, 1_000_000_000_000) # Conditional expression

True

Which of the following numbers is in the range `range(0, 1000000000000, 3)`:
- 27_541_111
- 27_541_112
- 27_541_113

In [36]:
27_541_113 in range(0, 1000000000000, 3)

True

In [38]:
for x in range(0,5):
    print(x**2)

0
1
4
9
16


In [ ]:
n = 10
for x in range(n):
    print('hi, ', x)

### _Quick overview of `string` and `list`_

In [45]:
myname = "dostoevsky"
myname

'dostoevsky'

In [40]:
type(myname)

str

In [46]:
len(myname) #length

10

In [48]:
# indexing
print(myname[0])
print(myname[1])
print(myname[9])
#print(myname[10])
print(myname[11])


d
o
y


IndexError: string index out of range

### Iterating over a string

In [50]:
# Iterating over a string
name = input('Enter your name:')

for j in name:
    print(j)
    sleep(.3)


Enter your name: iftach


i
f
t
a
c
h


### Iterating over a list

In [51]:
mylist = [1, 2.0, 'three', 'a', 'b', 'cee']
len(mylist)

6

In [52]:
from time import sleep
# Iterating over a list
mylist = [1, 2.0, 'three', 'a', 'b', 'cee']

for x in mylist:
    print(x, '-->', type(x))
    sleep(.3)

1 --> <class 'int'>
2.0 --> <class 'float'>
three --> <class 'str'>
a --> <class 'str'>
b --> <class 'str'>
cee --> <class 'str'>


### Using `enumerate()` with `for`

First, let's look at some examples of what `enumerate` does.

In [ ]:
e = enumerate('dostoevsky')
type(e)

In [ ]:
list(range(10))

In [ ]:
print(e)
list(e)

In [ ]:
e = enumerate('dostoevsky')
list(e)

In [ ]:
len('dostoevsky')

In [ ]:
e = enumerate('dostoevsky')

for counter, each_char in e:
    if counter % 2 == 1:
        print(each_char)


In [ ]:
e = enumerate('dostoevsky')

In [ ]:
print(type(e))

In [ ]:
next(e)

In [ ]:
e = enumerate(range(0,1_000_000_000_000,5))
type(e)

In [ ]:
next(e)

Now, let's use it in a `for` loop

In [ ]:
my_list = ['a', 5, 'c', 10, 'ff', '55']
for i, each_item in enumerate(my_list):
    print('Current number index is: ', i, '.     And current item is: ', each_item)

#### Why the emphasis on `range`
The *python* way vs. the "generic" way

In [ ]:
days_of_week = ['sunday', 'monday', 'tuesday', 'wednesday', 'thursday', 'friday']

In [ ]:
# Generic way
for i in range(len(days_of_week)):
    print(days_of_week[i])

In [ ]:
# Python way / R
for each_day in days_of_week:
    print(each_day)

## Example of using loops

In [ ]:
n_days = 30
r0 = 1.5
number_of_infected = 1
for each_day in range(0, n_days):
    
    # update how many are infected?
    number_of_infected = number_of_infected * r0
    
    print('Day ' + str(i) + ', number of infected = ' + str(round(number_of_infected,1)))

In [ ]:
import matplotlib.pyplot as plt

n_days = 100
r0 = 1.5
number_of_infected = [1]

for each_day in range(1, n_days):
    # update how many are infected?
    number_of_infected.append(number_of_infected[each_day-1] * r0)
    
plt.plot(list(range(0,n_days)), number_of_infected)
plt.xlabel('Days since first infection')
plt.ylabel('Number of people infected')
plt.show()

In [ ]:
# INITIATE LOCK-DOWN
rate_of_infection = .90
day_counter = n_days

# Count how many days until number of infected goes below 20,000
while number_of_infected[day_counter-1] > 20_000:
    
    # update how many are infected?
    number_of_infected.append(number_of_infected[day_counter-1] * rate_of_infection)
    
    # update days counter
    day_counter = day_counter + 1

# How many days until we end lockdown?
print(str(day_counter) + ' days until lockdown ends')

plt.plot(list(range(0,day_counter)), number_of_infected)
plt.xlabel('Days since first infection')
plt.ylabel('Number of people infected')
plt.show()

## `break` and `continue` in loops

In [53]:
mylist = ['a', 'b', 'cc', 'd', 'ee', 'ffffff']
print(mylist)

['a', 'b', 'cc', 'd', 'ee', 'ffffff']


In [54]:
print(mylist[0], len(mylist[0]))
print(mylist[5], len(mylist[5]))

a 1
ffffff 6


In [56]:
mylist

['a', 'b', 'cc', 'd', 'ee', 'ffffff']

In [55]:

for each_item in mylist:
    if len(each_item) > 1:
        continue
    
    print(each_item)

print('The last item copied from the list (each_time) = ' + each_item)

a
b
d
The last item copied from the list (each_time) = ffffff


In [ ]:
each_item

In [57]:
mylist = ['a', 'b', 'cc', 'd', 'ee', 'fffffff']
for each_item in mylist:
    if len(each_item) > 1:
        break
    
    print(each_item)

print('The last item copied from the list (each_time) = ' + each_item)

a
b
The last item copied from the list (each_time) = cc


In [ ]:
for index in range(1, 20):
    if index % 5 == 0:
        break
    
    print(index)

print('END OF PROGRAM')

In [58]:
# Write a loop that asks user to enter a password until password is correct
password = 'abc123'
while True:
    user_input = input('Enter password:')
    if user_input == password:
        break

print('Correct password')

Enter password: 123
Enter password: jgjgjg
Enter password: nvnvnv
Enter password: abc123


Correct password


## `for` exercises

Write a program that displays a temperature conversion table for degrees Celsius and
degrees Fahrenheit. The table should include rows for all temperatures between 0
and 100 degrees Celsius that are multiples of 10 degrees Celsius. Include appropriate
headings on your columns.